### ISSUE TITLE
Remote storage and cache - Huge amount of (useless ?) cache

### ISSUE THREAD

**Issue by Nuranto** [magento/magento2#35820]
### Summary (*)

We have L2 caching enabled. But I guess you can reproduce the issue with just setting up the cache backend to Filesystem.
When a S3 remote system is configured, a file cache is created for each single image (with tags `flysystem` and `mage`).

The result is that we crashed one of our server by launching `bin/magento catalog:images:resize --async` and `bin/magento queue:consumers:start media.storage.catalog.image.resize`, because we reached server's inode limits.

### Proposed solution

I'm not sure if this cache is relevant, and it should probably be removed.
If I'm wrong and that cache is usefull for performance, then we should disable that caching on bulk operation like `bin/magento catalog:images:resize`

### Additional Information
We have tried to debug this in detail and for me, the cache path is `var/cache`:

<img width="1431" alt="image" src="https://user-images.githubusercontent.com/51681618/196215405-ca5a8047-7085-49ef-b394-3e55ea300b49.png">

I have also observed that in order to save the cache it is using the below method from [colinmollenhour](https://github.com/colinmollenhour/Cm_Cache_Backend_File)
/
[Cm_Cache_Backend_File](https://github.com/colinmollenhour/Cm_Cache_Backend_File):

https://github.com/colinmollenhour/Cm_Cache_Backend_File/blob/034bf73adfdc5b02057ae3ef2a2255b381f46944/File.php#L178-L205

And the `$hash` variable is empty in this method.

This function is calling when we have tried to debug the below command:

`XDEBUG_CONFIG=idekey=PHPSTORM bin/magento remote-storage:sync`

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.4/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users with no workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [X] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by Nuranto** [comment]
**Additional informations**

We tried to disable L2 cache, and use our dedicated-to-cache redis instead.

Here is what I can see after the `image.resize` is done : 
```
127.0.0.1:6379> info memory
# Memory
used_memory:68741172840
used_memory_human:64.02G
```

64G of cache, for only 50K products and 229 869 entries in `catalog_product_entity_media_gallery`. With resizes, category and cms, multiple store views, we have a total of 5 663 932 files in our S3 instance.

I guess that's not a good thing for performances especially since in our case, we use Minio as S3 remote engine which is in the same cluster, so a call to redis is probably not significantly faster than requesting Minio directly.

**Comment by engcom-Hotel** [comment]
Hello @Nuranto,

Thanks for the report and collaboration!

We have tried to reproduce the issue in Magento 2.4-develop but the issue is not reproducible for us. We have verified it in the `var/cache` folder.

We have followed the below steps:

- Install the fresh Magento instance
- Install Sample Data
- Run the command `bin/magento catalog:images:resize --async`

Please let us know if we have missed anything.

Thanks

**Comment by Nuranto** [comment]
Hello @engcom-Hotel ,

Have you enabled & configured a remote storage (S3) ?

**Comment by engcom-Hotel** [comment]
Hello @Nuranto,

Yes, we have tried it with S3 enabled. But unable to reproduce the issue. Please refer to the below screenshot:

<img width="841" alt="image" src="https://user-images.githubusercontent.com/51681618/186922394-ee67f49b-c3ec-49f3-822b-319e4d12cc91.png">

The below path doesn't have any changes:

`var/cache`

Please let us know if we have missed anything.

Thanks

**Comment by Nuranto** [comment]
Hello @engcom-Hotel,

Have you run the consumer ? `bin/magento queue:consumers:start media.storage.catalog.image.resize`

I think you can also run the resize in synchronous mode, it should gave the same result.

**Comment by engcom-Hotel** [comment]
Hello @Nuranto,

Yes, we tried it with both Sync and Async commands. But the issue was not reproducible for us. Can you please suggest the path where this cache has been saved for you?

Thanks

**Comment by Nuranto** [comment]
Well, in my case I can see all that cache files with : `ls -l /dev/shm/* |grep JPG`

Each of this files look like this : 
```
a:4:{s:4:"hash";s:0:"";s:5:"mtime";i:1661872995;s:6:"expire";i:1661880195;s:4:"tags";s:0:"";}
{"media\/catalog\/product\/w\/e\/we***.jpg":{"path":"media\/catalog\/product\/w\/e\/we***.jpg","dirname":"media\/catalog\/product\/w\/e","basename":"we***","extension":"jpg","filename":"we***","type":"file","size":86308,"timestamp":1646762270,"visibility":null,"mimetype":"image\/jpeg","extra":{"image-height":"600","image-width":"450"}}}
```

You can find the cache logic in `Magento\RemoteStorage\Driver\Adapter\CachedAdapter`, for example have a look at `fileExists` method, I guess that's where it gets generated.
See also `Magento\RemoteStorage\Driver\Adapter\Cache\Generic::updateMetadata` where you can see that a cache file is created on a image filepath basis. 

However I don't understand why you cannot reproduce this. The `$persist` parameter of `updateMetadata` is hardcoded to true in `fileExists` so it should not be related to a configuration...

**Comment by engcom-Hotel** [comment]
Hello @Nuranto,

I have tried to debug this in detail and for me, the cache path is `var/cache`:

<img width="1431" alt="image" src="https://user-images.githubusercontent.com/51681618/196215405-ca5a8047-7085-49ef-b394-3e55ea300b49.png">

I have also observed that in order to save the cache it is using the below method from [colinmollenhour](https://github.com/colinmollenhour/Cm_Cache_Backend_File)
/
[Cm_Cache_Backend_File](https://github.com/colinmollenhour/Cm_Cache_Backend_File):

https://github.com/colinmollenhour/Cm_Cache_Backend_File/blob/034bf73adfdc5b02057ae3ef2a2255b381f46944/File.php#L178-L205

And the `$hash` variable is empty in this method.

This function is calling when we have tried to debug the below command:

`XDEBUG_CONFIG=idekey=PHPSTORM bin/magento remote-storage:sync`

Need your inputs on these findings.

Thanks

**Comment by Nuranto** [comment]
Yes, in my case, as specified in my first message, we are using L2 caching, with basic configuration took from the docs, with this line : 
```
'local_backend_options' => [
            'cache_dir' => '/dev/shm/'
        ],
```
But I don't think the cache location matters for this issue.



For the `$hash` being empty, I'm also not sure if this is relevant. In my previous message, you can see that this data is also an empty string : `s:4:"hash";s:0:""`.

But if you're looking there, it means you managed to reproduce the caching behaviour of 1 cache file per image ?

**Comment by engcom-Hotel** [comment]
Hello @Nuranto,

Thanks for the reply!

Yes I can manage to reproduce the issue that the caching behavior of 1 cache file per image by running the below command:

`bin/magento remote-storage:sync`

But as I told you that this has been due to [colinmollenhour](https://github.com/colinmollenhour/Cm_Cache_Backend_File)
/
[Cm_Cache_Backend_File](https://github.com/colinmollenhour/Cm_Cache_Backend_File) repo. But we can prevent this behavior by passing `false` in the below function call:

https://github.com/magento/magento2/blob/6057eb8a22f310ff354f1bc9dc95d2e68f1298ae/app/code/Magento/RemoteStorage/Driver/Adapter/CachedAdapter.php#L161

Please suggest.

Thanks

**Comment by Nuranto** [comment]
Hello @engcom-Hotel ,

I don't understand why you think the issue is in `colinmollenhour`. The redis extension does just what it is asked for : save a cache entry for each file .

But I agree with your fix proposal : we should not persist this cache.

**Comment by engcom-Hotel** [comment]
Hello @Nuranto,

Thanks for your reply!

As per the process, we are moving this ticket for a fix. Hence confirming the issue.

Thanks

**Comment by Nuranto** [comment]
Hello @engcom-Hotel,

Do you have any news on this (on Jira's issue ?). Is there a patch available or planned ?

This issue causes stability issues on my production environment (after a while, it saturates available RAM which provokes nodes crash). So if you tell me that the ticket is still pending on Jira's side, I'll work on a PR.
